### PR TITLE
Add ability to view soft clipped bases in alignments track 

### PR DESCRIPTION
**Pull request by peterkxie**
Addresses #602. Option to show soft clipped bases added to track menu called 'Show Soft Clipping'. This option can be toggled on and off and is default off. The option does not persist upon refresh currently (is a volatile state). 

When turned on for alignments track, the soft clipped bases will be drawn on each read. The colors for the bases are lightened to differentiate them from the regular mismatch bases. Any base outside A C T G (i.e N) will be drawn as black to be quickly noticeable. There is a maxClippingSize config option that users can edit if their file has soft clip lengths that exceed the default value (10,000 bp)

### CODE REVIEW

**Comment by peterkxie** [comment]
sure! let me try that out

**Comment by peterkxie** [comment]
> Adding `softClip` to the types in ServerSideRenderer and BoxRenderer probably isn't the right thing to do, because it's super specific to the alignments stuff. A better approach would might be to `extend` those interfaces in PileupRenderer to make types with more parameters, and override the `cachedLayoutIsValid` method in PileupRenderer. Could you try that?

Should the same be done to the sortObject? Since that is specific to Pileup currently too. Or will sort be expanded to other tracks and can be left in BoxRendererType

**Comment by rbuels** [comment]
Would be good to do it with sortObject too, yeah

File: packages/alignments/src/PileupRenderer/PileupRenderer.ts [hunk 1]
```diff
@@ -37,17 +46,79 @@ interface LayoutRecord {
...
+  sortObject: unknown
+  showSoftClip: unknown
+}
+
+class SoftClipLayoutSession extends LayoutSession {
+  sortObject: unknown
+
+  showSoftClip: unknown
+
+  cachedLayoutIsValid(cachedLayout: CachedPileupLayout) {
```
Comment (cmdcolin) [inline packages/alignments/src/PileupRenderer/PileupRenderer.ts:68]: This should probably use super.cachedLayoutIsValid(cachedLayout) + the logic specific to this

File: packages/alignments/src/PileupRenderer/PileupRenderer.ts [hunk 2]
```diff
@@ -37,17 +46,79 @@ interface LayoutRecord {
...
+  sortObject: unknown
+  showSoftClip: unknown
+}
+
+interface CachedPileupLayout extends CachedLayout {
+  sortObject: unknown
+  showSoftClip: unknown
+}
+
+class SoftClipLayoutSession extends LayoutSession {
```
Comment (cmdcolin) [inline packages/alignments/src/PileupRenderer/PileupRenderer.ts:63]: Rename to just PileupLayoutSession

File: packages/alignments/src/PileupTrack/model.ts [hunk 3]
```diff
@@ -118,7 +118,8 @@ export default (
           ...self.composedRenderProps,
           ...getParentRenderProps(self),
           trackModel: self,
-          sortObject: self.sortObject,
+          sortObject: this.sortObject,
+          showSoftClip: getParentRenderProps(self).trackModel.showSoftClipping,
```
Comment (cmdcolin) [inline packages/alignments/src/PileupTrack/model.ts:122]: at some point, the PileupTrack should really take control of this variable and not have to look to it's parent for this IMO, but I wouldn't call it a blocker here.

**Comment by cmdcolin** [comment]
I think the flakyness of the imagesnapshot test may be due to non-determinism in selecting a given block to snapshot..we might want to key on region in data-testid so that we get the exact block we want to snapshot

**Comment by cmdcolin** [comment]
:clap:

**Comment by rbuels** [comment]
Yay!

**Pull request merged**